### PR TITLE
BUG: Merge with empty dataframe may raise IndexError

### LINKED ISSUE #10824: BUG: Merge with empty DataFrame raise IndexError
```
import numpy as np
import pandas as pd

df1 = pd.DataFrame([], columns=['a', 'b', 'c'])
df2 = pd.DataFrame(np.random.randn(3, 3), columns=['x', 'y', 'z'])

# OK
pd.merge(df1, df2, how='right', left_index=True, right_index=True)
#      a    b    c         x         y         z
# 0  NaN  NaN  NaN  0.665359  0.087728 -0.608138
# 1  NaN  NaN  NaN -0.730847  0.882151  0.175648
# 2  NaN  NaN  NaN  2.370834 -1.347337 -0.478547

pd.merge(df1, df2, how='outer', left_index=True, right_index=True)
#      a    b    c         x         y         z
# 0  NaN  NaN  NaN  0.665359  0.087728 -0.608138
# 1  NaN  NaN  NaN -0.730847  0.882151  0.175648
# 2  NaN  NaN  NaN  2.370834 -1.347337 -0.478547

# NG, this should be the same as above
pd.merge(df1, df2, how='right', left_on='a', right_index=True)
# IndexError: cannot do a non-empty take from an empty axes.

pd.merge(df1, df2, how='outer', left_on='a', right_index=True)
# IndexError: cannot do a non-empty take from an empty axes.
```

This works if both ``DataFrame`` are empty.

```
df1 = pd.DataFrame([], columns=['a', 'b', 'c'])
df2 = pd.DataFrame([], columns=['x', 'y', 'z'])

pd.merge(df1, df2, how='right', left_on='a', right_index=True)
# Empty DataFrame
# Columns: [a, b, c, x, y, z]
# Index: []

pd.merge(df1, df2, how='outer', left_on='a', right_index=True)
# Empty DataFrame
# Columns: [a, b, c, x, y, z]
# Index: []
```

### PR DESCRIPTION
**Pull request by sinhrks**
Closes #10824.

### CODE REVIEW

File: pandas/tools/merge.py [hunk 1]
```diff
@@ -273,9 +273,17 @@ def _get_join_info(self):
...
+                else:
+                    join_index = self.right.index.take(right_indexer)
+                    left_indexer = np.array([np.nan] * len(join_index))
             elif self.left_index:
-                join_index = self.right.index.take(right_indexer)
+                if len(self.right) > 0:
+                    join_index = self.right.index.take(right_indexer)
+                else:
+                    join_index = self.left.index.take(left_indexer)
+                    right_indexer = np.array([np.nan] * len(join_index))
```
Comment (jreback) [inline pandas/tools/merge.py]: the indxers should be ``np.array([-1])*len(join(index))`` no?

File: pandas/tools/tests/test_merge.py [hunk 2]
```diff
@@ -737,6 +737,83 @@ def test_left_merge_empty_dataframe(self):
...
+                      dict(left_index=True, right_on='x'),
+                      dict(left_on='a', right_index=True),
+                      dict(left_on='a', right_on='x')]:
+
+            result = pd.merge(left, right, how='inner', **kwarg)
+            tm.assert_frame_equal(result, exp_in, check_dtype=False)
+            result = pd.merge(left, right, how='right', **kwarg)
+            tm.assert_frame_equal(result, exp_in, check_dtype=False)
+
+            print(kwarg)
```
Comment (jreback) [inline pandas/tools/tests/test_merge.py]: pls remove prints

File: pandas/tools/merge.py (hunk 1 above)
Comment (sinhrks) [inline pandas/tools/merge.py]: It is used for filling when ``left_on`` is specified. When use ``np.array([-1])*len(join(index))`` or leave it as it is, the result will be:

```
import pandas as pd
left = pd.DataFrame([], columns=['a', 'b', 'c'])
right = pd.DataFrame([[1, 2, 3], [4, 5, 6], [7, 8, 9]],
                             columns=['x', 'y', 'z'])
result = pd.merge(left, right, how='right', left_on='a', right_index=True)
result
#    a    b    c  x  y  z
# 0  0  NaN  NaN  1  2  3
# 1  1  NaN  NaN  4  5  6
# 2  2  NaN  NaN  7  8  9
```

File: pandas/tools/tests/test_merge.py (hunk 2 above)
Comment (sinhrks) [inline pandas/tools/tests/test_merge.py]: Thanks, removed.

File: pandas/tools/merge.py (hunk 1 above)
Comment (jreback) [inline pandas/tools/merge.py]: these MUST be ``int64`` (and the len's happen to be 0).

Comment (jreback) [inline pandas/tools/merge.py]: further pls examine the returned dtypes (they are ``object``), not 100% sure that is right

Comment (sinhrks) [inline pandas/tools/merge.py]: OK, change indexer to ``int (-1)``, and added ``dtype`` checks (done by ``assert_frame_equal`` by default).

**Comment by jreback** [comment]
lgtm

**Pull request merged**